contracts/src/Assets.sol

```diff
@@ -58,34 +50,35 @@ library Assets {
         address token,
         address sender,
         ParaID destinationChain,
-        address destinationAddress,
+        MultiAddress calldata destinationAddress,
         uint128 amount
     ) external returns (bytes memory payload, uint256 extraFee) {
         AssetsStorage.Layout storage $ = AssetsStorage.layout();
-        if (destinationChain == assetHubParaID) {
-            // AssetHub parachain doesn't support Ethereum-style addresses
-            revert InvalidDestination();
-        }
 
         _transferToAgent(assetHubAgent, token, sender, amount);
 
-        payload = SubstrateTypes.SendToken(address(this), token, destinationChain, destinationAddress, amount);
-        extraFee = $.sendTokenFee;
-
-        emit IGateway.TokenSent(sender, token, destinationChain, abi.encodePacked(destinationAddress), amount);
-    }
-
-    /// @dev transfer tokens from the sender to the specified
-    function _transferToAgent(address assetHubAgent, address token, address sender, uint128 amount) internal {
-        if (!token.isContract()) {
-            revert InvalidToken();
-        }
-
-        if (amount == 0) {
-            revert InvalidAmount();
+        if (destinationChain == assetHubParaID) {
+            if (destinationAddress.isAddress32()) {
+                payload = SubstrateTypes.SendTokenToAssetHubAddress32(token, destinationAddress.asAddress32(), amount);
+            } else {
+                revert Unsupported();
+            }
+        } else {
+            if (destinationAddress.isAddress32()) {
+                payload = SubstrateTypes.SendTokenToAddress32(
+                    token, destinationChain, destinationAddress.asAddress32(), amount
+                );
+            } else if (destinationAddress.isAddress20()) {
+                payload = SubstrateTypes.SendTokenToAddress20(
+                    token, destinationChain, destinationAddress.asAddress20(), amount
+                );
+            } else {
+                revert Unsupported();
+            }
```

Comment:
What if the `destinationAddress` is `Index`?

Reply:
That's not supported right now, requires more plumbing on the BridgeHub side, to convert the message into the right XCM.

Not many parachains use `pallet_indices` anyway.